pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh

```diff
@@ -21,18 +21,29 @@ fi
 # use a copy of patchelf.
 LD_LIBRARY_PATH=$out/lib $LD_BINARY $out/bin/cp $out/bin/patchelf .
 
+# make a copy of /lib so we don't try patchelf'ing a library that
+# patchelf itself links against -- this may cause segfaults due to
+# lazy loading of binary images from disk
+LD_LIBRARY_PATH=$out/lib $LD_BINARY $out/bin/cp -r $out/lib lib
+
 for i in $out/bin/* $out/libexec/gcc/*/*/*; do
     if [ -L "$i" ]; then continue; fi
     if [ -z "${i##*/liblto*}" ]; then continue; fi
     echo patching "$i"
-    LD_LIBRARY_PATH=$out/lib $LD_BINARY \
+    LD_LIBRARY_PATH=lib $LD_BINARY \
         ./patchelf --set-interpreter $LD_BINARY --set-rpath $out/lib --force-rpath "$i"
 done
 
-for i in $out/lib/librt-*.so $out/lib/libpcre*; do
```

Comment:
The line you highlighted was adding an RPATH to `librt.so`, which does not have an RPATH.

The unexplained reason why this worked is that every binary in the bootstrap-files is linked to `librt.so` (either directly or indirectly).  So every binary in the bootstrap-files would end up loading `librt.so`, and the RPATH from `librt.so` would therefore get included in the "aggregate RPATH" for the binary.  This is why we were getting away with leaving so many `nuke-refs`'d references to `/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-...` laying around.

If `librt.so` gets patched, and if the unstated assumption that every binary is linked to it (directly or indirectly), then I think that `patchelf`'ing `libpcre.so` is unnecessary.  I can check this in the morning if knowing would be useful.

In general the reason why the current `unpack-bootstrap-tools.sh`'s use of `patchelf` works involves a lot of unexplained reasoning like this.  It works, but it is really hard to understand *why* it works.

The main benefit of this PR is to make the reasoning really simple and straightforward: we replace all of the `/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-...` references created by `nuke-refs` (with one exception: `libgcc_s.so`, which is explained in [this comment](https://github.com/NixOS/nixpkgs/blob/a9e0e1ea21675baddaac63ee1887b9763cd39702/pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh#L40)), and do nothing else.

If it would help, I could rewrite the script to not only check that the `.so` has an RPATH before `patchelf`ing it, but also checks that the RPATH is `/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-...`.  This would bring us even closer to the goal of "undo what `nuke-refs` did and do nothing else".

Reply:
Thanks for the explanation, I'm starting to make sense of it a bit. I think in general this does sound like the right direction to go.

Further improvements can always be done in follow-up PRs as well.